config/src/main/kotlin/org/springframework/security/config/annotation/web/AuthorizeHttpRequestsDsl.kt

```diff
@@ -0,0 +1,253 @@
+     * using the pattern as an ant pattern will be used.
+     *
+     * @param pattern the pattern to match incoming requests against.
+     * @param access the [AuthorizationManager] to secure the matching request
+     * (i.e. created via hasAuthority("ROLE_USER"))
+     */
+    fun authorize(pattern: String,
+                  access: AuthorizationManager<RequestAuthorizationContext> = authenticated) {
+        authorizationRules.add(
+            PatternAuthorizationManagerRule(
+                pattern = pattern,
+                patternType = PATTERN_TYPE,
+                rule = access
+            )
+        )
+    }
+
+    /**
+     * Adds a request authorization rule for an endpoint matching the provided
+     * pattern.
+     * If Spring MVC is on the classpath, it will use an MVC matcher.
+     * If Spring MVC is not on the classpath, it will use an ant matcher.
+     * The MVC will use the same rules that Spring MVC uses for matching.
+     * For example, often times a mapping of the path "/path" will match on
+     * "/path", "/path/", "/path.html", etc.
+     * If the current request will not be processed by Spring MVC, a reasonable default
+     * using the pattern as an ant pattern will be used.
+     *
+     * @param method the HTTP method to match the income requests against.
+     * @param pattern the pattern to match incoming requests against.
+     * @param access the [AuthorizationManager] to secure the matching request
+     * (i.e. created via hasAuthority("ROLE_USER"))
+     */
+    fun authorize(method: HttpMethod,
+                  pattern: String,
+                  access: AuthorizationManager<RequestAuthorizationContext> = authenticated) {
+        authorizationRules.add(
+            PatternAuthorizationManagerRule(
+                pattern = pattern,
+                patternType = PATTERN_TYPE,
+                httpMethod = method,
+                rule = access
+            )
+        )
+    }
+
+    /**
+     * Adds a request authorization rule for an endpoint matching the provided
+     * pattern.
+     * If Spring MVC is on the classpath, it will use an MVC matcher.
+     * If Spring MVC is not on the classpath, it will use an ant matcher.
+     * The MVC will use the same rules that Spring MVC uses for matching.
+     * For example, often times a mapping of the path "/path" will match on
+     * "/path", "/path/", "/path.html", etc.
+     * If the current request will not be processed by Spring MVC, a reasonable default
+     * using the pattern as an ant pattern will be used.
+     *
+     * @param pattern the pattern to match incoming requests against.
+     * @param servletPath the servlet path to match incoming requests against. This
+     * only applies when using an MVC pattern matcher.
+     * @param access the [AuthorizationManager] to secure the matching request
+     * (i.e. created via hasAuthority("ROLE_USER"))
+     */
+    fun authorize(pattern: String,
```

Comment:
Regarding your concern about method matching, you might need to do the overloading yourself, that is, without default parameters. You can have a method that takes two Strings and then invokes an overloaded method that takes two Strings and an AuthorizationManager.

Reply:
Thanks for the review, @jzheaux!
I analyzed this problem a bit more thoroughly.
1. There is some inconsistency between lambda config and Kotlin DSL for `authorizeRequests`
    - In lambda config, if endpoint matching is specified, a user must specify the access
    - Kotlin DSL config allows to omit the access, then "authenticated" will be applied by default
    ```
    // lambda config
    http.authorizeRequests { authorize ->
        authorize.mvcMatchers("/hello").authenticated()  // user must explicitly specify access
    }
    
    // Kotlin DSL
    http {
        authorizeRequets {
            authorize("/hello")  // user can omit the access, then "authenticated" is applied implicitly
        }
    }
    ```
2. Problem in my case (rather migration problem than method overloading)
The case when a problem may appear is a user has `authorizeRequests` config with SpEL in Kotlin DSL
```
authorizeRequests {
    authorize("/user/{userName}", "#userName == 'user'")
}
```
and migrates to new API simply adding Http to the method name
```
authorizeHttpRequests {
    authorize("/user/{userName}", "#userName == 'user'")
}
```
The code compiles successfully and the user may not pay attention that his/her SpEL expression does not configure access anymore.
If the app's API is covered by tests, then the user will get the error and no problem here.
Intellij also shows a hint that the second parameter is changed to servletPath.

If I remove "authenticated" default in this method:
```
fun authorize(pattern: String,
              servletPath: String,
              access: AuthorizationManager<RequestAuthorizationContext>)
```
Then the compiler shows an error, and forces the user pay attention that his/her SpEL config is not valid anymore.

But removing for this particular method adds a special case to the framework API.
So, I think the choice is to leave default "authenticated" config (app tests or Intellij hint solution) or remove "authenticated" by default for all methods in Kotlin DSL for `authorizeHttpRequests` (compilation error solution).

3. I realized that the test case with path variable value can be convered the following config instead of SpEL:
```
@Bean
open fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
    val access = AuthorizationManager { _: Supplier<Authentication>, context: RequestAuthorizationContext ->
        AuthorizationDecision(context.variables["userName"] == "user")
    }
    http {
        authorizeHttpRequests {
            authorize("/user/{userName}", access)
        }
    }
    return http.build()
}
```

Reply:
Another option is omitting the `servletPath` from this method entirely. 
That's the approach we use for `authorizeRequests`. 
The `servletPath` only applies to MVC matchers, so it could be confusing in the Kotlin DSL because the pattern could end up being used in an Ant matcher or MVC matcher.

Reply:
I'm not sure if I've got it right, but `authorizeRequests` uses `servletPath`.
If an Ant matcher is used, then `servletPath` is ignored:
```
 when (rule.patternType) {
       PatternType.ANT -> requests.antMatchers(rule.httpMethod, rule.pattern).access(rule.rule)
       PatternType.MVC -> requests.mvcMatchers(rule.httpMethod, rule.pattern)
                .apply { if(rule.servletPath != null) servletPath(rule.servletPath) }
                .access(rule.rule)
 }
 ```
 Removing the `servletPath` from the method in `authorizeHttpRequests` makes it different from `authorizeRequests` and a bit reduces the API.<br>
 I can propose a solution for the migration problem by using an explicit type for `servletPath` parameter (new 4th option).
 Thus, the `authorize` methods of `authorizeHttpRequests` provide the same functionality and defaults as `authorizeRequests`, and a user will get a compilation error during migration.
 The API looks like:
 ```
 authorizeHttpRequests {
     authorize("/path", servletPath("/spring"), hasRole("ADMIN"))
 }
 ```
 Please, correct if I've got smth wrong in the option with omitting the `servletPath` parameter.

Reply:
Thanks @ysavchen. You are correct about the `servletPath`.
You've also made me see another issue with the `authorizeRequests` DSL, for which I created the issue gh-10957.

I think your suggestion to remove "authenticated" by default for all methods in Kotlin DSL for `authorizeHttpRequests` (compilation error solution) is the best approach.

Reply:
Thanks @eleftherias 
Updated PR